### ISSUE TITLE
[Tracking Issue] Oopsie Woopsie (doesn't even compile, and if it did...)

### ISSUE THREAD

**Issue by wbrickner** [wbrickner/const_cge#2]
Made a significant fucky-wucky and jumped the gun on releasing this package.

The whole point is to allow you to deploy network anywhere without thinking about the target.

On microcontrollers (the juiciest spot to deploy EANT2 / CGE models), you often have no floating point functions like `tanh` and `sqrt` etc.  `micromath` provides **approximations** to these functions, but they're often really bad (sqrt(2) == 1.5).  Plus, I'd have to lock in users to `micromath`, and we also seem to have a `libm` port available.

This should be fine, if only you _train_ your network using poor approximations over these functions.
However, this will break the fundamental promises of the package, as networks may behave very differently on no_std targets, and to train using approximations you _need to know in advance_ where your network will end up, and it cannot easily be shared (I want small utility networks to literally be crates someday).

Possible solutions, all bad:

- **Lawful evil**: copy the code that implements the `std`'s floating point approximations of elementary functions mostly verbatim.  These are probably giant lookup tables or big fat taylor series or something unpleasant, or maybe theyre compiler built-ins, I've never looked.  They're pretty fast on desktop so maybe they're not so bad, idk.

- **Chaotic evil**: allow for micromath's approximations to produce compounding errors inside networks, warn the user, hope for the best. Supply the `tanh` approximation which it lacks.  Perhaps we can examine the range over which these approximations are relevant and improve them, making a `micromath` support tailored for neural network evaluation. Idk, lots of work.

- Or, we can abandon floating point support all together.  I think this option sucks too.

I think the correct solution is going to be to examine how bad these compounding errors actually get, and possibly provide custom shim that takes over for micromath.

**Comment by wbrickner** [comment]
All crate components yanked until I can figure this out & provide a reasonable solution.
Unclear to me if `libm` approximations are any good / what their properties are.

**Comment by wbrickner** [comment]
- Several of the activation functions should actually be branchless bit hacks on all platforms & perform no actual floating point math (linear, threshold, sign, ReLU).  I should tell users that if they want their model to be guaranteed to function identically on all targets, then use one of these activations.

- The differences emerge in the sigmoid, tanh, soft_sign, and bent_identity activations.

- I think a reasonable default would be to depend on `libm` for `no_std` targets, as I think it mostly matches identically with existing `std` code, with the option to switch to `micromath` approximations if that's what they prefer.  `micromath` has no `tanh`, so I can either steal `libm`'s `tanh` or just use another approximation in place.

**Comment by wbrickner** [comment]
Ok so I went down the rabbit hole on designing this correctly, mostly to support my vision of "netcrates" while also solving the numerical backend problem.  This was really hard.  The trick is to only allow netcrates to expose networks as data, not code, and allow code to be generated by the end user from the data. But the data is not actually available at macro expansion time! The way we do this: 
- a proc-macro that loads the data from file into a literal,
- produces a `macro_rules!` macro that _when invoked_ will create an _invocation_ of the main `const_cge` proc macro.
- This is sinful and terrible and confusing but it works magic and avoids the dependency hell that would exist if networks were published as generated code.

The new solution:
- Supports for the end user any floating point backend that `core_cge` offers, and this can be upgraded in a centralized way.
- Does not require netcrate authors to understand floating point backends or forward specific features to `libm`, `micromath`, `std`, etc.  This would end up as hell.
- netcrate authors only see:
```rust
const_cge::netcrate!("path/to/awesomenet.cge");
```

- netcrate consumers only see:
```rust
use awesome_net;
#[network(awesome_net)]
struct AwesomeNet;
```
and then `const_cge` end users must pick their backend in features like:
```toml
const_cge = { version = "0.1.3", features = ["libm"] }
const_cge = { version = "0.1.3", default-features = false, features = ["std"] }
const_cge = { version = "0.1.3", default-features = false, features = ["micromath"] }
```

This might limit some composability in netcrates, but we can fix that.

- Say you want to provide a tiny pretrained neural turing machine, and you want to provide the user a nice implementation to use it. My preferred solution is to publish /another macro/ (sorry)
```rust
macro_rules! ntm {
  ($name: ident) => {
    impl $name {
      const MY_CUSTOM_CONST: usize = ...;
      fn my_custom_extension(&mut self) -> { ... }
    }
  }
}
pub(crate) use ntm; // expose the extension macro
```

And an end user desiring this extension:
```rust
#[network(ntm_zoo::some_task, numeric_type = f32)]
struct SomeTask;
ntm!(SomeTask);
```

Not elegant, but it solves some really hard problems, and allows for a maintainable and flexible ecosystem for embedded ML.

**Comment by wbrickner** [comment]
Resolved, as far as I can tell anyway.
